### ISSUE TITLE
VoiceOver screen reader support

### ISSUE THREAD

**Issue by slook** [bretth18/seeleseek#55]
**Is your feature request related to a problem? Please describe.**

A blind Soulseek user reports that they are having multiple accessibility problems accessing the network because all of the mainstream clients (QT, Nicotine+, slskd) have been tried but each of them have various issues that cause them to not work well enough without the aid of a visual display. See original post on Reddit, hyperlink:

https://www.reddit.com/r/Soulseek/comments/1utbazc/totally_blind_putting_out_a_call_for_macos/

For example, it may be possible to setup an account and establish a connection but it is very difficult or impossible to read the search results or configure shared folders, due to the existence of accessibility bugs or lack of particular features on certain widget implementations that are required by screen reading software.

It is not possible for them to operate many legacy user interfaces with a screen, because they do not use a screen to operate their computer (due to being "totally blind") and they have to use other sensory inputs (such as sound and haptic feedback for example) instead of vision.

It has been suggested that Swift is an ideal framework because the OP has found it from their expertise to have "good accessibility tools built-in". They do not yet know about the [seeleseek](https://seeleseek.net/) project, which has an interface that is implemented with Swift.

**Describe the solution you'd like**

The user wishes to make contact with a Swift developer who is able to make a user-friendly Soulseek client that works, or can be made to work, properly with [VoiceOver](http://www.apple.com/accessibility/voiceover/) on macOS so that blind people (and perhaps people with other kinds of disabilities) can fully participate on the network.

In return for ongoing technical support that may be required from time to time (as accessibility bugs are inevitability going to be discovered by them), the OP is willing to provide ongoing constructive feedback to the project such as proactively testing development builds on their system and making helpful bug reports to assist in the design, development, refinement and maintenance of accessibility features.

I'm guessing the starting point will be for the user to perform testing on the accessibility of the installation procedure, which may in and of itself require some degree of thought and preparation or remediation, before the feasibility of basic application functions as to determining which parts of the application, if any, are usable.

Therefore, it could be prudent to perform some cursory testing to eliminate any fundamental blockers before deploying the application onto the end-user's own testing environment. Thereafter, the OP would be best placed to explain exactly what modifications may be required according their specific runtime needs.

For some generic guidance about how software developers can best approach doing this work, see hyperlinks:
https://developer.apple.com/documentation/accessibility/performing-accessibility-testing-for-your-app
https://developer.apple.com/documentation/uikit/supporting-voiceover-in-your-app

**Describe alternatives you've considered**

- **_SoulseekQT_**
Doesn't work very well with screen readers at all, due to having minimal accessibility compliance, difficult to install and runs badly on macOS. It is unlikely that the official client will ever meet the special accessibility needs of blind users.

- **_Nicotine+_**
The current state of the GTK toolkit has poor accessibility, especially in the TreeView and ColumnView widgets which are central to its interface, and it is not known when this will improve upstream. We do not work with nor use macOS ourselves so testing generally on the target platform in minimal, and our knowledge of VoiceOver internals is practically non-existent. As such we are unable to meet the user's needs at this time.

- **_slskd_**
This is an advanced home-server application that's unsuited for every-day use as a desktop client, and while it's web interface is partially operable it does not appear to be compliant enough to do critical tasks, like navigating between modes and setting up shared folders (that even fully-sighted users have significant difficulties with!). An entirely reworked user interface would probably be required to meet the accessibility needs of blind users.

- **_Alternative/proprietary screen reader_**
It would be preferable to utilize accessibility tooling that is already built-in to the operating system rather than resorting to elaborate workarounds or expensive solutions for interfacing with non-compliant applications.

- **_Personal assistant/support worker_**
While personal assistance with computing tasks may be available to the user sometimes (such as in educational or vocational settings), it is the user's fundamental human right to be able to configure and use their private home computer, including any software applications they may choose to install on it, freely and independently.

- **_Alternative P2P network_**
No other network provides the functionality and community that exists on Soulseek.

**Additional context**

The user is technically competent and wishes to help with testing and development to improve accessibility compliance in a third-party client until they are able to successfully operate all of its essential and advanced functions without using a visual display.

Implementing accessibility compliance benefits all kinds of users, but identifying issues and fixing accessibility bugs can only be done properly by collaborating with people who actually have a special need for using accessibility tooling on their system. Opportunities for getting feedback from blind testers come up rarely so this would be a good chance get real-world feedback from an enthusiastic user who is willing and able to aid UX design and development of the project's modern front-end GUI.

They do not yet know about the [seeleseek](https://seeleseek.net/) project. I cannot personally inform them because I'm not on Reddit, although @mathiascode has responded to briefly indicate that the [Nicotine+](https://nicotine-plus.org/) project is unfortunately not yet suitable for their use case on the target system. In my opinion the [seeleseek](https://seeleseek.net/) project is much better placed for this than any other client, since it has the right kind of platform resources to take the lead on this issue for the time being.

The user can be contacted by responding to their post on Reddit to direct towards this issue, at link:
https://www.reddit.com/r/Soulseek/comments/1utbazc/totally_blind_putting_out_a_call_for_macos/

**Comment by bretth18** [comment]
Thank you for the detailed issue and for surfacing this to me!

I'm happy to take this on and agree with everything you've stated here. Will reach out to the OP on reddit as well.

**Comment by slook** [comment]
Have you heard anything back from the OP yet or notified them about the new release?

**Comment by bretth18** [comment]
> Have you heard anything back from the OP yet or notified them about the new release?

last correspondence with the OP was on July 27th, but they said they'd downloaded and had been using the first release that contained the VoiceOver support changes. think it's safe to call this one done + closed
